src/clients/node/src/node.zig

```diff
@@ -15,783 +19,470 @@ const CreateTransfersResult = tb.CreateTransfersResult;
+    allocated: u16,
+    completed: u16,
+};
 
-    if (context.client.messages_available == 0) {
-        translate.throw(
-            env,
-            "Too many outstanding requests - message pool exhausted.",
-        ) catch return null;
+fn on_completion(
+    completion_ctx: usize,
+    client: tb.tb_client_t,
+    packet: *tb.tb_packet_t,
+    result_ptr: ?[*]const u8,
+    result_len: u32,
+) callconv(.C) void {
+    switch (packet.status) {
+        .ok => {},
+        .too_much_data => unreachable, // We limit packet data size during request().
+        .invalid_operation => unreachable, // We check the operation during request().
+        .invalid_data_size => unreachable, // We set correct data size during request().
     }
-    const message = context.client.get_message();
-    errdefer context.client.release(message);
-
-    const operation = @as(Operation, @enumFromInt(@as(u8, @intCast(operation_int))));
-    const body_length = decode_events(
-        env,
-        argv[2],
-        operation,
-        message.buffer[@sizeOf(Header)..],
-    ) catch |err| switch (err) {
-        error.ExceptionThrown => return null,
-    };
-
-    // This will create a reference (in V8) to the user's JS callback that we must eventually also
-    // free in order to avoid a leak. We therefore do this last to ensure we cannot fail after
-    // taking this reference.
-    const user_data = translate.user_data_from_value(env, argv[3]) catch return null;
-    context.client.request(@as(u128, @bitCast(user_data)), on_result, operation, message, body_length);
-
-    return null;
-}
 
-/// The batch has already been encoded into a byte slice. This means that we only have to do one
-/// copy directly into an available message. No validation of the encoded data is performed except
-/// that it will fit into the message buffer.
-fn raw_request(env: c.napi_env, info: c.napi_callback_info) callconv(.C) c.napi_value {
-    var argc: usize = 4;
-    var argv: [4]c.napi_value = undefined;
-    if (c.napi_get_cb_info(env, info, &argc, &argv, null, null) != c.napi_ok) {
-        translate.throw(env, "Failed to get args.") catch return null;
+    const results = result_ptr.?[0..result_len];
+    switch (@as(Operation, @enumFromInt(packet.operation))) {
+        inline else => |op| {
+            const event_count = @divExact(packet.data_size, @sizeOf(StateMachine.Event(op)));
+            const buffer: BufferType(op) = .{
+                .ptr = @ptrCast(packet.data.?),
+                .count = event_count,
+            };
+
+            const Result = StateMachine.Result(op);
+            const completed: []const Result = @alignCast(std.mem.bytesAsSlice(Result, results));
+            @memcpy(buffer.results()[0..completed.len], completed);
+
+            packet.data_size = @bitCast(BufferSize{
+                .allocated = @intCast(event_count),
+                .completed = @intCast(completed.len),
+            });
+        },
```

Comment:
Suggestion: no need for taking an alias for `results` and avoid introducing new names such as `completed`;

```suggestion
    switch (@as(Operation, @enumFromInt(packet.operation))) {
        inline else => |op| {
            const event_count = @divExact(packet.data_size, @sizeOf(StateMachine.Event(op)));
            const buffer: BufferType(op) = .{
                .ptr = @ptrCast(packet.data.?),
                .count = event_count,
            };

            const Result = StateMachine.Result(op);
            const results: []const Result = @alignCast(
                std.mem.bytesAsSlice(Result, result_ptr.?[0..result_len]),
            );
            @memcpy(buffer.results()[0..results.len], results);

            packet.data_size = @bitCast(BufferSize{
                .allocated = @intCast(event_count),
                .completed = @intCast(results.len),
            });
        },
```